lib/Runtime/Library/JavascriptSet.cpp

```diff
@@ -6,379 +6,665 @@
+        this->Add(value);
+        return;
 
-        return scriptContext->GetLibrary()->CreateSetIterator(set, JavascriptSetIteratorKind::Value);
+    case SetKind::ComplexVarSet:
+        this->AddToComplexVarSet(value);
+        return;
+
+    default:
+        Assume(UNREACHED);
     }
+}
 
-    Var JavascriptSet::EntryGetterSymbolSpecies(RecyclableObject* function, CallInfo callInfo, ...)
+void JavascriptSet::Clear()
+{
+    JS_REENTRANCY_LOCK(jsReentLock, this->GetScriptContext()->GetThreadContext());
+    this->list.Clear();
+    switch (this->kind)
     {
-        ARGUMENTS(args, callInfo);
-
-        Assert(args.Info.Count > 0);
+    case SetKind::EmptySet:
+        return;
+    case SetKind::IntSet:
+        this->u.intSet->ClearAll();
+        return;
+    case SetKind::SimpleVarSet:
+        this->u.simpleVarSet->Clear();
+        return;
+    case SetKind::ComplexVarSet:
+        this->u.complexVarSet->Clear();
+        return;
+    default:
+        Assume(UNREACHED);
+    }
+}
 
-        return args[0];
+bool
+JavascriptSet::IsInIntSet(Var value)
+{
+    Var taggedInt = JavascriptConversion::TryCanonicalizeAsTaggedInt<false>(value);
+    if (!taggedInt)
+    {
+        return false;
     }
+    int32 intVal = TaggedInt::ToInt32(taggedInt);
+    return this->u.intSet->Test(intVal);
+}
 
-    void JavascriptSet::Add(Var value)
+template <bool isComplex>
+bool
+JavascriptSet::DeleteFromVarSet(Var value)
+{
+    SetDataNode * node = nullptr;
+    if (isComplex
+        ? !this->u.complexVarSet->TryGetValueAndRemove(value, &node)
+        : !this->u.simpleVarSet->TryGetValueAndRemove(value, &node))
     {
-        if (!set->ContainsKey(value))
-        {
-            SetDataNode* node = list.Append(value, GetScriptContext()->GetRecycler());
-            set->Add(value, node);
-        }
+        return false;
     }
 
-    void JavascriptSet::Clear()
+    this->list.Remove(node);
+    return true;
+}
+
+bool
+JavascriptSet::DeleteFromSimpleVarSet(Var value)
+{
+    Var simpleVar = JavascriptConversion::TryCanonicalizeAsSimpleVar(value);
+    if (!simpleVar)
     {
-        // TODO: (Consider) Should we clear the set here and leave it as large as it has grown, or
-        // toss it away and create a new empty set, letting it grow as needed?
-        list.Clear();
-        set->Clear();
+        return false;
     }
 
-    bool JavascriptSet::Delete(Var value)
+    return this->DeleteFromVarSet<false>(simpleVar);
+}
+
+bool JavascriptSet::Delete(Var value)
+{
+    JS_REENTRANCY_LOCK(jsReentLock, this->GetScriptContext()->GetThreadContext());
+    switch (this->kind)
     {
-        if (set->ContainsKey(value))
+    case SetKind::EmptySet:
+        return false;
+
+    case SetKind::IntSet:
+        if (!IsInIntSet(value))
         {
-            SetDataNode* node = set->Item(value);
-            list.Remove(node);
-            return set->Remove(value);
+            return false;
         }
+        this->PromoteToSimpleVarSet();
```

Comment:
>this->PromoteToSimpleVarSet(); [](start = 8, length = 30)

Why is a conversion required?

Reply:
it's hard to maintain list after deleting elements from an int set, because we don't have pointer to the element in the list. there might be some more complex work that can allow us to push the conversion until point where we need an iterator, but it seemed like pretty big complication, and scenario i'm looking at doesn't use delete (it just uses clear, which is fine).
I think this deserves a comment for now, and if it comes up, maybe we can improve this in future

---
In reply to: [174327605](https://github.com/Microsoft/ChakraCore/pull/4816#discussion_r174327605) [](ancestors = 174327605)